app-emulation/wa-linux-agent/files/0001-flatcar-changes.patch

```diff
@@ -0,0 +1,229 @@
+ from .default import DefaultOSUtil
++from .flatcar import FlatcarUtil
+ from .freebsd import FreeBSDOSUtil
+ from .gaia import GaiaOSUtil
+ from .iosxe import IosxeOSUtil
+@@ -82,6 +83,8 @@ def _get_osutil(distro_name, distro_code_name, distro_version, distro_full_name)
+         return DebianOSBaseUtil()
+ 
+     if distro_name in ("flatcar", "coreos") or distro_code_name in ("flatcar", "coreos"):
++        if Version(distro_version) > Version("3530"):
++            return FlatcarUtil()
+         return CoreOSUtil()
+ 
+     if distro_name in ("suse", "sle_hpc", "sles", "opensuse"):
+diff --git a/azurelinuxagent/common/osutil/flatcar.py b/azurelinuxagent/common/osutil/flatcar.py
+new file mode 100644
+index 00000000..a34fb7dc
+--- /dev/null
++++ b/azurelinuxagent/common/osutil/flatcar.py
+@@ -0,0 +1,37 @@
++#
++# Copyright 2023 Microsoft Corporation
++#
++# Licensed under the Apache License, Version 2.0 (the "License");
++# you may not use this file except in compliance with the License.
++# You may obtain a copy of the License at
++#
++#     http://www.apache.org/licenses/LICENSE-2.0
++#
++# Unless required by applicable law or agreed to in writing, software
++# distributed under the License is distributed on an "AS IS" BASIS,
++# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
++# See the License for the specific language governing permissions and
++# limitations under the License.
++#
++# Requires Python 2.6+ and Openssl 1.0+
++#
++
++import os
++import shutil
++
++import azurelinuxagent.common.conf as conf
++
++from azurelinuxagent.common.osutil.coreoscommon import CoreosCommonUtil
++
++
++class FlatcarUtil(CoreosCommonUtil):
++
++    def conf_sshd(self, disable_password):
++        # make sure that the config file stops being a symlink
++        conf_file_path = conf.get_sshd_conf_file_path()
++        conf_file_path2 = f"{conf_file_path}.wal.tmp"
++        shutil.copy(conf_file_path, conf_file_path2)
++        os.remove(conf_file_path)
++        os.rename(conf_file_path2, conf_file_path)
++        super(CoreosCommonUtil, self).conf_sshd(disable_password)
++        pass
+diff --git a/setup.py b/setup.py
+index d38d74d6..eaf19194 100755
+--- a/setup.py
++++ b/setup.py
+@@ -125,12 +125,16 @@ def get_data_files(name, version, fullname):  # pylint: disable=R0912
+                           src=["init/arch/waagent.service"])
+     elif name in ('coreos', 'flatcar'):
+         set_bin_files(data_files, dest=agent_bin_path)
+-        set_conf_files(data_files, dest="/usr/share/oem",
+-                       src=["config/coreos/waagent.conf"])
+         set_logrotate_files(data_files)
+         set_udev_files(data_files)
+-        set_files(data_files, dest="/usr/share/oem",
+-                  src=["init/coreos/cloud-config.yml"])
++        if int(version.split('.')[0]) > 3530:
++            set_conf_files(data_files, dest="/etc",
++                           src=["config/coreos/waagent.conf"])
```

Comment:
The reason this file is not under `/usr` as part of the sysext images is that this file needs to be writable, or?

Reply:
You know, this patch is something I'd like to eventually take to upstream. But I was not really sure what direction the changes there should take. Should setup.py in this case be only for us, the distro maintainers, or should it be usable also for people wanting to install a newer version of WALinuxAgent on their Flatcar instances. I'm not sure if these two use cases are reconcilable. I'm leaning towards the former, so maybe I should update the patch to install waagent.conf to `/usr/share/flatcar/etc`?

Reply:
So, in the end I'm installing this file in `/usr/share/waagent/waagent.conf`. and in the `waagent.service` unit file I have an `ExecStartPre` line which creates a symlink in `/etc/waagent.conf` that points to `/usr/share/waagent/waagent.conf`.